ipa-core/src/protocol/ipa_prf/malicious_security/verifier.rs

```diff
@@ -10,6 +10,43 @@ use crate::{
     },
 };
 
+/// This function computes the shares that sum to zero from the zero-knowledge proofs.
+pub fn compute_g_differences<F, const P: usize, const λ: usize>(
+    zkp: &Vec<[F; P]>,
+    challenges: &[F],
+) -> Vec<F>
+where
+    F: PrimeField,
+{
+    // compute denominator
+    let lagrange_denominator: CanonicalLagrangeDenominator<F, P> =
+        CanonicalLagrangeDenominator::<F, P>::new();
+
+    // compute g_r with "empty" first spot
+    let g_r = iter::once(F::ZERO)
+        .chain(
+            challenges
+                .iter()
+                .zip(zkp)
+                .map(|(challenge, zkp)| interpolate_at_r(zkp, *challenge, &lagrange_denominator)),
+        )
+        .collect::<Vec<_>>();
+
+    // compute g_sum)
+    let g_sum = zkp
+        .iter()
+        .map(compute_sum_share::<F, λ, P>)
+        // append spot for final sum
+        .chain(iter::once(F::ZERO))
+        .collect::<Vec<_>>();
+
+    g_sum
+        .iter()
+        .zip(g_r)
+        .map(|(g_sum, g_r)| *g_sum - g_r)
+        .collect()
+}
+
 ///
 /// Distributed Zero Knowledge Proofs algorithm drawn from
 /// `https://eprint.iacr.org/2023/909.pdf`
```

Comment:
I would rather not do this here, but instead do it within the function itself.  Basically things the check for the first stage that it equals -m/2.